Arc 23: Sharing Application information on the compiled application's bytes

SudoWeezy opened the pull request:
This is a new version of the [PR Create arc-23: Append the hash of the contract description to the compiled application's bytes #80](https://github.com/algorandfoundation/ARCs/pull/80) 

It is based on this [comment](https://github.com/algorandfoundation/ARCs/pull/80#issuecomment-1372318550).

This ARC used IPFS CID to store information about an application.

SudoWeezy committed:
first commit ARC-23 with ipfs

SudoWeezy committed:
Arc-23 using CID

fabrice102 commented on ARCs/arc-0023.md:
```diff
@@ -0,0 +1,76 @@
+eg:`bafybeihdzhfifv46p27ee6weqq6h7ydttzwni7gvdogr3mg6rwscdyzici/Abi.json`.
+
+### Appending
+The encoded byte is appended to the compiled application as a <a href="https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes">bytecblock</a> containing 1-byte constant which is the encoded object.
```
```suggestion
The encoded byte is appended to the compiled application as a <a href="https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes">bytecblock</a> containing `[]byte` constant which is the encoded object.
```

fabrice102 replied:
There are 3 properties we want of the way the information is included in the program:

1. Easy to extract: it should be easy to detect when ARC-23 information are added and extract them. In particular, this should not require the ability to fully parse TEAL code, nor find boundaries of op-codes. The number of false positive should be negligible and the code to parse should be easy enough.
2. Easy to securely add to programs: it should be easy to add to new programs in a fully automated and secure way. This means that adding it should not break security. In addition, the TEAL compiler should be able to handle this gracefully. See also https://github.com/algorandfoundation/ARCs/pull/80#issuecomment-1252662213
3. Short: it should not be too long.

If we enforce that the CID is always a V1 CID w/ SHA256, this makes it easier because then

```
bytecblock b64(YXJjMjMBcBIg48nKgteefr5CesSEPH/gc55s1HzVG40dsN6NpCHjKBI=)
```

is always encoded as (in hexadecimal):

`0x2601296172633233` followed by 36 bytes of the CID

(`0x26` is the `bytecblock` opcode, `0x01` is the number of []byte strings, `0x29` is the length = 41 bytes, `0x6172633233` is hexadecimal of `arc23)

If in addition it's always supposed to be at the end of the TEAL code, then it's very easy to parse:
go to 44 bytes before end of file.

Now, an alternative would be:

```
bytecblock "arc23" b64(CID)
```

is easier to parse in case the CID may have different length since then you can look for `0x2602056172633233` as prefix that is then followed by a varuint of the length of the CID and the CID.
This is easier than concatenating "arc23" to the CID and having a single []byte bytecblock element because we have a very easy long prefix (otherwise, it can get messy).

But if the length of the CID is fixed, then the first option saves 1 byte.

fabrice102 commented on ARCs/arc-0023.md:
```diff
@@ -0,0 +1,76 @@
+The <a href="https://github.com/multiformats/cid">CID</a> allows to access the corresponding files using <a href="https://docs.ipfs.tech/">IPFS</a>.
+
+The CID **MUST** represent one file or a folder of files.
+The minimum requirement for this ARC is a file representing the ABI (ABI as described in [ARC-4](arc-0004.md))
```
```suggestion
The minimum requirement for this ARC is a folder containing a file `contract.json` representing the contract metadata (as described in the ABI in [ARC-4](arc-0004.md), and as extended by future potential ARCs).
```

fabrice102 replied:
The above comment is to add the ability to extend the contract metadata.
Also I think ABI is a general term defining methods, ...
while the specific metadata can be called "contract metadata" and named "contract.json".

fabrice102 replied:
This also follows convention there https://github.com/algorand-devrel/beaker/blob/master/examples/account_storage/DiskHungry.artifacts/contract.json

fabrice102 commented on ARCs/arc-0023.md:
```diff
@@ -0,0 +1,76 @@
+### CID
+The <a href="https://github.com/multiformats/cid">CID</a> allows to access the corresponding files using <a href="https://docs.ipfs.tech/">IPFS</a>.
+
+The CID **MUST** represent one file or a folder of files.
```
In addition, CID can have multiple format and hash algortihms.
I would recommend enforcing these.

```suggestion
The CID **MUST**:
* represent a folder of files
* be a version V1 CID
* use SHA-256 hash algorithm
```

fabrice102 commented on ARCs/arc-0023.md:
```diff
@@ -0,0 +1,76 @@
+Otherwise, if there is more than one file, the CID must correspond to a folder of files. The ABI **MUST** be accessible by querying and IPFS node with `{cid}/Abi.json`
+eg:`bafybeihdzhfifv46p27ee6weqq6h7ydttzwni7gvdogr3mg6rwscdyzici/Abi.json`.
+
+### Appending
```
```suggestion
### Including the encoded information to the program
```

fabrice102 replied:
Also one question is whether the appending should be done at the end of the clear program and/or approval program.
Since clear programs are less space constrained in general, maybe we should put them there? What do you think?

fabrice102 commented on ARCs/arc-0023.md:
```diff
@@ -0,0 +1,76 @@
+The size of the compiled application + the bytecblock should be, at most, the maximum size of a compiled application according to the latest consensus parameters supported by the compiler.
+
+## Rationale
+By appending the IPFS CID of the folder containing information about the Application, any user with access to the blockchain could easily verify the Application and the ABI of the Application and interact with it.
```
```suggestion
By appending the IPFS CID of the folder containing information about the Application, any user with access to the blockchain could easily verify the Application and the contract metadata of the Application and interact with it.
```

fabrice102 commented on ARCs/arc-0023.md:
```diff
@@ -0,0 +1,76 @@
+The encoded byte is appended to the compiled application as a <a href="https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes">bytecblock</a> containing 1-byte constant which is the encoded object.
+> The reason to use `bytecblock` is that adding a `bytecblock` opcode at the end of a TEAL program does not change the semantics of the program, as long as opcodes are correctly aligned, there is no jump after the last position (that would make the program fail without `bytecblock`), and there is enough space left to add the opcode.
+The size of the compiled application + the bytecblock should be, at most, the maximum size of a compiled application according to the latest consensus parameters supported by the compiler.
+
```
MAybe add a paragraph about non-deployed application.

### Use in development

In development, it may be preferable to not publish the application information on IPFS.
Tools that need these information (such as local block explorer) can request this information as a .zip or .tar.gz containing at the root all the required files.

fabrice102 reviewed (commented):
I made various detailed comments but here are the high-level bits:

* clarifying the exact CID requirement (folder vs file, V1, SHA256)
* proposing to compress the encoded byte string by using CID bytes and removing information that is redundant
* clarifying the bytecblock encoding
* proposing naming convention for files that was used in beaker

SudoWeezy committed:
Update ARCs/arc-0023.md

Co-authored-by: Fabrice Benhamouda <fabrice+githubfabrice102@algorand.foundation>

SudoWeezy committed:
Apply suggestions from code review

Co-authored-by: Fabrice Benhamouda <fabrice+githubfabrice102@algorand.foundation>

SudoWeezy committed:
Update using bytes to reduce size.

SudoWeezy committed:
Sentence for ARC extension

SudoWeezy committed:
rename Application.py -> application.py

fabrice102 committed:
Various changes to ARC-23

* require IPFS CAR for production to ensure proper archiving
  and add explicit guidance
* add more step-by-step examples for each stage
* make naming more consistent
* make the Python script PIP-8 and add comments/functions
* remove the use of non-folder
* clarify where the bytecblock may be as well as risks related to new TEAL versions
* rename "Folder" into "application_information" in assets
  and ensure __pycache__ is ignored for reproducible results
* require explicitly use of approval.teal

fabrice102 commented:
I've made a PR with some additional various changes to be discussed:
https://github.com/algorandfoundation/ARCs/pull/159

Also changed to standards track instead of meta.

fabrice102 committed:
Fix arcw issues

fabrice102 committed:
minor

SudoWeezy committed:
Merge pull request #159 from fabrice102/ARC-23

ARC-23: misc changes to draft in #157

SudoWeezy committed:
Last call before pushing inside main repo

ori-shem-tov commented on ARCs/arc-0023.md:
```diff
@@ -0,0 +1,186 @@
+> At least with TEAL v8 and before, appending the `bytecblock` to the end of the program should add exactly 44 bytes (1 byte for opcode `bytecblock`, 1 byte for 0x01 -the number of byte strings-, 1 byte for 0x29 the length of the encoded information byte string, 41 byte for the encodedin information byte string)
+
+The `bytecblock` **MAY** be placed anywhere in the TEAL source code as long as it does not modify the semantic of the TEAL source code.
+However, if `approval.teal` is provided as an application information file, the `bytecblock` **SHOULD** be the last opcode of the TEAL program.
```
```suggestion
However, if `approval.teal` is provided as an application information file, the `bytecblock` **SHOULD** be the last opcode of the deployed TEAL program.
```

SudoWeezy committed:
TEAL -> deployed TEAL

Co-authored-by: ori-shem-tov <45538125+ori-shem-tov@users.noreply.github.com>

SudoWeezy committed:
Adding file versioning for python file
